docs: derek's first refactor of the quickstart

leeederek opened the pull request:
## Description:
This is a minor refactor of the quickstart. This change aims to make a few sections a bit more clear, and emphasizes some of the learnings we want new users to take away from each section.

fixes #304 

## Is this change user facing?
YES
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful information. -->

Commit:
docs: derek refactor of the quickstart, v1.1

Commit:
fix: Make setup instructions more clear

leeederek commented on docs/docs/quickstart.md:
```diff
@@ -162,16 +166,18 @@ UUID           Name       Ports                                                S
+We call this approach [multi-phase runs][multi-phase-runs-reference]. While multi-phase runs has powerful benefits over traditional scripting, it also means _you cannot reference Execution values like IP address in Starlark_ because they simply don't exist at Interpretation time. We'll explore how Kurtosis gracefully handles values generated during the Execution phase at the Interpretation phase  ...
 
-We call this approach [multi-phase runs][multi-phase-runs-reference]. While it has powerful benefits, the stumbling point for new Kurtosis users is that _you cannot reference Execution values like IP address in Starlark_ because they simply don't exist at Interpretation time. We'll see how to work around this limitation later.
+**This section introduced Kurtosis' ability to validate environment definitions work as intended, _before_ they're run, helping developers catch errors sooner and save resources when configuring multi-container test environments.**
```
```suggestion
**This section introduced Kurtosis' ability to validate that definitions work as intended, _before_ they are run - helping developers catch errors sooner & save resources when configuring multi-container test environments.**
```

leeederek committed:
fix typo

leeederek commented on docs/docs/quickstart.md:
```diff
@@ -182,17 +188,19 @@ Our two options for seeding a Postgres database are:
 
-Normally seeding a database would require downloading the seed data to your machine, starting Postgres, and writing a pile of Bash to copy the seed data to the Postgres server and run a `pg_restore`. If you forgot to check if the database is available, you may get flakes when you try to use the seeding logic in a test. 
+##### Without Kurtosis
+Normally going this route seeding requires downloading the seed data to your local machine, starting Postgres, and writing a pile of Bash to copy the seed data to the Postgres server, and then finally running a `pg_restore` command. If you forgot to check if the database is available, you may get flakes when you try to use the seeding logic in a test. 
```
```suggestion
Normally going this route (using `pg_restore`) requires downloading the seed data to your local machine, starting Postgres, writing a pile of Bash to copy the seed data to the Postgres server, and then finally running the `pg_restore` command. If you forgot to check if the database is available, you may get flakes when you try to use the seeding logic in a test. 
```

leeederek committed:
fix another typo

leeederek committed:
fix files artifact uuid/name mismatch

leeederek committed:
make some titles more prominent

leeederek committed:
consolidate the conclusion

leeederek committed:
tidy up conclusion again

galenmarchetti commented on docs/docs/quickstart.md:
```diff
@@ -7,7 +7,7 @@ slug: /quickstart
 ------------
 
-Welcome to the [Kurtosis][homepage] quickstart! This guide will take ~15 minutes and will walk you through building a basic Kurtosis package.
+Welcome to the [Kurtosis][homepage] quickstart! This guide will take ~15 minutes and will walk you through configuring and building a basic Kurtosis package that can be used as an environment for running multi-container tests.
```
I would leave this section as-is, just to keep brevity

leeederek replied:
```suggestion
Welcome to the [Kurtosis][homepage] quickstart! This guide will take ~15 minutes and will walk you through building a basic Kurtosis package.
```

galenmarchetti commented on docs/docs/quickstart.md:
```diff
@@ -31,7 +31,9 @@ There are many ways to get help and give feedback. First, every Kurtosis command
 ## Setup
 
-Before you proceed, make sure you have [Kurtosis installed][installing-kurtosis-guide] (or [upgraded to latest][upgrading-kurtosis-guide] if you already have it), Docker is started, and the Docker engine is running.
+Before you proceed, please make sure you have:
```
this is great

galenmarchetti committed:
Update docs/docs/quickstart.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -136,14 +140,14 @@ def run(plan, args):
 
 :::info
-This clean-and-run process will be your dev loop for the rest of the quickstart.
+This "clean-and-run" process will be your dev loop for the rest of the quickstart as we add more services and operations to our distributed application.
```
nice; this is better

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -95,17 +97,19 @@ UUID   Name   Ports   Status
+In this section, we created a `main.star` file that prints `Hello, world`. The `.star` extension corresponds to [the Starlark language developed at Google][starlark-github-repo], a dialect of Python for configuring the [Bazel build system][bazel-github]. [Kurtosis uses Starlark for the same purpose of configuring builds][starlark-explanation], except that we're building a distributed application  ...
 
-When you ran the Starlark, you got `Created enclave: quickstart`. An [enclave][enclaves-explanation] is a Kurtosis primitive that can be thought of as an *ephemeral test environment*, on top of Docker or Kubernetes, for a distributed application. The distributed applications that you define with Starlark will run inside enclaves. 
+When you ran `main.star`, you got `Created enclave: quickstart`. An [enclave][enclaves-explanation] is a Kurtosis primitive that can be thought of as an *ephemeral test environment*, on top of Docker or Kubernetes, for a distributed application. The distributed applications that you define with Starlark will run inside enclaves. If you'd like, you can tear down your enclave and any of their artif ...
```
Nice

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -95,17 +97,19 @@ UUID   Name   Ports   Status
 
 Enclaves are intended to be easy to create, easy to destroy, cheap to run, and isolated from each other. Use enclaves liberally!
 
+**In this section, you've experienced just how quickly and effortlessly Kurtosis can be used to generate and tear down ephemeral test environments on top of Docker or Kubernetes.**
```
Did we see that though? Feels too early to say that.

leeederek replied:
A few reasons why I think this is valid and should stay:
1. The user _did_ in fact see this: they wrote some Starlark and ran that code inside an isolated container and we showed the user how to (optionally) tear down their enclave using `kurtosis clean`. It's not a huge "wow" moment or something revolutionary, but we used this section to introduce the user to how quickly + easily it was to get an ephemeral environment up and running
2. For consistency's sake, every other section had a 1-2 sentence conclusion recapping a Kurtosis ability.

mieubrisse replied:
MY PROBLEM: it feels like it's reaching too far to say anything about environments given that they've ONLY printed "Hello, world". They haven't truly spun up an environment, nor tore it down, nor actually done anything iwth containers at all.

leeederek replied:
Got it. You've convinced me. 

```suggestion
```

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -182,17 +188,19 @@ Our two options for seeding a Postgres database are:
+Both are possible in Kurtosis, but for this tutorial we'll use `pg_restore` to seed our database with a TAR of DVD rental information, [courtesy of postgresqltutorial.com](https://www.postgresqltutorial.com/postgresql-getting-started/postgresql-sample-database/). 
 
-Normally seeding a database would require downloading the seed data to your machine, starting Postgres, and writing a pile of Bash to copy the seed data to the Postgres server and run a `pg_restore`. If you forgot to check if the database is available, you may get flakes when you try to use the seeding logic in a test. 
+##### Without Kurtosis
```
Does this level of header actually work?

leeederek replied:
yeah, its really small though. I think its the same font size as **bolded** font, but with extra padding below and above. See below: 
<img width="605" alt="image" src="https://user-images.githubusercontent.com/103802618/228271669-b29f2a5b-93c0-4b17-8792-ae1368922349.png">

mieubrisse replied:
Would prefer bolded text then, because headers go into GH's autogenerated table of contents

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -182,17 +188,19 @@ Our two options for seeding a Postgres database are:
+Normally going this route (using `pg_restore`) requires downloading the seed data to your local machine, starting Postgres, writing a pile of Bash to copy the seed data to the Postgres server, and then finally running the `pg_restore` command. If you forgot to check if the database is available, you may get flakes when you try to use the seeding logic in a test. 
 
-You could try Docker Compose to volume-mount the data TAR into the Postgres server, but you'd still need to handle Postgres availability and sequencing the `pg_restore` afterwards.
+Alternatively, you could use Docker Compose to volume-mount the data TAR into the Postgres server, but you'd still need to handle Postgres availability and sequencing the `pg_restore` afterwards.
```
cool

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -361,49 +369,70 @@ psql -U app_user -d app_db -c '\dt'
+### Review: Add some data
+So what just happened?
+
+#### We created a Kurtosis package
```
this is better (same note to check header levels though)

leeederek replied:
<img width="566" alt="image" src="https://user-images.githubusercontent.com/103802618/228323747-79303cd0-a6d2-4030-bd11-877d51970273.png">

Here's a screenshot of the rendered text on my preview branch: what do you think?

mieubrisse commented on docs/docs/quickstart.md:
````diff
@@ -361,49 +369,70 @@ psql -U app_user -d app_db -c '\dt'
 ```
 
-...created a dependency on [the external Kurtosis package living here][data-package-example]. 
+A [files artifact][files-artifacts-reference] is Kurtosis' first-class data primitive and is a TGZ of arbitrary files living inside an enclave. So long as a files artifact exists, Kurtosis knows how to mount its contents on a service.  
````
oooo nice, "Kurtosis' data primitive". I like this.

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -361,49 +369,70 @@ psql -U app_user -d app_db -c '\dt'
+A [files artifact][files-artifacts-reference] is Kurtosis' first-class data primitive and is a TGZ of arbitrary files living inside an enclave. So long as a files artifact exists, Kurtosis knows how to mount its contents on a service.  
 
-Your code then called that external dependency here...
+#### We mounting and seeded the data into our postgres instance
```
```suggestion
#### We mounted and seeded the data into our Postgres instance
```

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -361,49 +369,70 @@ psql -U app_user -d app_db -c '\dt'
 
-Your code then called that external dependency here...
+#### We mounting and seeded the data into our postgres instance
+Next, we mounted the seed data, stored in our enclave now as a files artifact, into our postgres instance using the `ServiceConfig.files` option:
```
s/postgres/Postgres/ (proper nouns for consistencyplease!)

leeederek replied:
```suggestion
Next, we mounted the seed data, stored in our enclave now as a files artifact, into our Postgres instance using the `ServiceConfig.files` option:
```

mieubrisse commented on docs/docs/quickstart.md:
````diff
@@ -586,15 +618,15 @@ def run(plan, args):
 ```
 
-In the line declaring the `postgres_url` variable in your `main.star` file, replace the `"postgres"` string with `POSTGRES_USER` to use the correct username we specified at the beginning of our file. Then rerun your dev loop:
+In the line declaring the `postgres_url` variable in your `main.star` file (line 58), replace the `"postgres",` string with `POSTGRES_USER,` to use the correct username we specified at the beginning of our file. Then re-run your dev loop:
````
I'd caution against line numbers, because they're likely not going to get updated in the future.

leeederek replied:
```suggestion
In the line declaring the `postgres_url` variable in your `main.star` file, replace the `"postgres",` string with `POSTGRES_USER,` to use the correct username we specified at the beginning of our file. Then re-run your dev loop:
```

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -618,10 +650,10 @@ ce90b471a982   postgres    postgres: 5432/tcp -> postgresql://127.0.0.1:59883
-### Review
-In this section, we declared a new PostgREST service (that we named `api` for readability) with a dependency on the Postgres service.
+### Review: Add an API
+In this section, we spun up new PostgREST service (that we named `api` for readability) with a dependency on the Postgres service. Normally, PostgREST needs to know the IP address or hostname of the Postgres service, and we said earlier that Starlark (the Interpretation phase) can never know Execution values. 
```
s/spun up new/spun up a new/

leeederek replied:
```suggestion
In this section, we spun up a new PostgREST service (that we named `api` for readability) with a dependency on the Postgres service. Normally, PostgREST needs to know the IP address or hostname of the Postgres service, and we said earlier that Starlark (the Interpretation phase) can never know Execution values. 
```

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -653,6 +685,8 @@ api = plan.add_service(
 
 ...Kurtosis simply swapped in the correct Postgres container Execution-time values. While future references take some getting used to, [we've found the feedback loop speedup to be very worth it][why-multi-phase-runs-explanation].
 
+**What you've just seen is Kurtosis' powerful ability grab to use dynamically generated data to setup service dependencies in multi-container test environments. We also saw how seamless it was to to run on-box CLI commands on a container.**
```
typo: "grab to use"

"setup" is a noun, but "set up" is a verb

leeederek replied:
```suggestion
**What you've just seen is Kurtosis' powerful ability to data generated at runtime to set up service dependencies in multi-container test environments. We also saw how seamless it was to to run on-box CLI commands on a container.**
```

mieubrisse commented on docs/docs/quickstart.md:
````diff
@@ -846,6 +880,9 @@ At a higher level, Kurtosis automatically deserialized the `[{"first_name":"Kevi
 def run(plan, args):
 ```
 
+**This section showed us how we can interact with the test environment we setup and also how to paramterize our distributed application for others future re-use.**
````
same: "setup" should be "set up"


"others future re-use"

leeederek replied:
```suggestion
**This section showed us how we can interact with the test environment we set up and also how to parametrize our distributed application for others to re-use.**
```

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -903,63 +940,27 @@ And that's it - you've written your very first distributed application in Kurtos
-- Added an API server
+- Started a Postgres database in an ephemeral, isolated test environment
+- Seeded your database by importing an external Starlark package from the internet
+- Setup an API server for our database and gracefully handled dynamically generated dependency data
```
"Setup" -> "set up"

leeederek replied:
```suggestion
- Set up an API server for our database and gracefully handled dynamically generated dependency data
```

mieubrisse reviewed (approved):
Nice man; I like your writing.

leeederek committed:
mounting --> mounted

Co-authored-by: Kevin Today <thetallmonkey@gmail.com>

leeederek committed:
setup --> set up

leeederek committed:
remove line numbers

leeederek committed:
fix typo

leeederek commented on docs/docs/quickstart.md:
```diff
@@ -361,49 +369,70 @@ psql -U app_user -d app_db -c '\dt'
 
-Your code then called that external dependency here...
+#### We mounted and seeded the data into our Postgres instance
+Next, we mounted the seed data, stored in our enclave now as a files artifact, into our postgres instance using the `ServiceConfig.files` option:
```
```suggestion
Next, we mounted the seed data, stored in our enclave now as a files artifact, into our Postgres instance using the `ServiceConfig.files` option:
```

leeederek committed:
postgres --> Postgres

leeederek committed:
typo

leeederek committed:
typos

leeederek commented:
@mieubrisse addressed most of your comments, but would still like to hear your thoughts on the remaining outstanding comments you left for me. Thanks!

mieubrisse commented on docs/docs/quickstart.md:
```diff
@@ -618,10 +650,10 @@ ce90b471a982   postgres    postgres: 5432/tcp -> postgresql://127.0.0.1:59883
+In this section, we spun up a new PostgREST service (that we named `api` for readability) with a dependency on the Postgres service. Normally, PostgREST needs to know the IP address or hostname of the Postgres service, and we said earlier that Starlark (the Interpretation phase) can never know Execution values. 
 
-Yet... PostgREST needs to know the IP address or hostname of the Postgres service, and we said earlier that Starlark (the Interpretation phase) can never know Execution values. How can this be?
+So how did the services get connected?
```
Looks like the question mark is getting deleted

leeederek committed:
downsize headers for with+w/o kurtosis

leeederek committed:
Merge branch 'main' into derek/refactor-quickstart-v1.1

leeederek committed:
remove hello-world concluding sentence

leeederek committed:
Merge branch 'main' into derek/refactor-quickstart-v1.1

leeederek commented on docs/docs/quickstart.md:
```diff
@@ -361,49 +368,70 @@ psql -U app_user -d app_db -c '\dt'
-But where did the data come from? 
-
-There are many ways to create files artifacts in an enclave. The simplest is to upload files from your local machine using [the `kurtosis files upload` command][kurtosis-files-upload-reference]. A more advanced way is to upload files using [the `upload_files` Starlark instruction][upload-files-reference] on the plan.
+This external Kurtosis package, named ["data-package"][data-package-example] contains the seed data for our postgres instance that we [referenced ealier](#add-some-data) as a `.tar` file.
```
```suggestion
This external Kurtosis package, named ["data-package"][data-package-example] contains the seed data for our Postgres instance that we [referenced earlier](#add-some-data) as a `.tar` file.
```

leeederek committed:
postgres --> Postgres & eailer --> earlier
